libs/cosmos-sdk/store/mpt/snapshot.go

```diff
@@ -0,0 +1,207 @@
+package mpt
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/state/snapshot"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+	mpttypes "github.com/okx/okbchain/libs/cosmos-sdk/store/mpt/types"
+)
+
+var (
+	gDisableSnapshot = false
+	gSnapshotRebuild = false
+)
+
+const (
+	// snapshotMemoryLayerCount snapshot memory layer count
+	// as we dont rollback transactions so we only keep 1 memory layer
+	snapshotMemoryLayerCount = 1
+)
+
+func DisableSnapshot() {
+	gDisableSnapshot = true
+}
+
+func SetSnapshotRebuild(rebuild bool) {
+	gSnapshotRebuild = rebuild
+}
+
+func (ms *MptStore) openSnapshot() error {
+	if ms == nil || ms.db == nil || ms.trie == nil || ms.db.TrieDB().DiskDB() == nil || gDisableSnapshot {
+		return fmt.Errorf("mpt store is nil or mpt trie is nil")
+	}
+	// If the chain was rewound past the snapshot persistent layer (causing
+	// a recovery block number to be persisted to disk), check if we're still
+	// in recovery mode and in that case, don't invalidate the snapshot on a
+	// head mismatch.
+	var recovery bool
+
+	version := ms.CurrentVersion()
+	if layer := rawdb.ReadSnapshotRecoveryNumber(ms.db.TrieDB().DiskDB()); layer != nil && *layer > uint64(version) {
+		if ms.logger != nil {
+			ms.logger.Error("Enabling snapshot recovery", "chainhead", version, "diskbase", *layer)
+		}
+		recovery = true
+	}
+	var err error
+	ms.snaps, err = snapshot.NewCustom(ms.db.TrieDB().DiskDB(), ms.db.TrieDB(), 256, ms.originalRoot, false, gSnapshotRebuild, recovery, ms.retriever)
```

Comment:
used the 'ms.db.TrieDB().DiskDB()' as a db maybe affect the statics?

Reply:
yes, and a later discussion is necessary

Reply:
after discussion the statistics can include snapshot's db operator